From e3fa43b4e3fd6b0db7fc7bfcc2db41093abe1377 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:25 -0400
Subject: [virt] kvm/ppc: book3s hv - Don't set DABR on POWER8

Message-id: <1410545655-205645-337-git-send-email-dzickus@redhat.com>
Patchwork-id: 94640
O-Subject: [RHEL7 PATCH 336/626] KVM: PPC: Book3S HV: Don't set DABR on POWER8
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit eee7ff9d2cc0eaaa00496bdf4193144104c7dc63
Author: Michael Neuling <mikey@neuling.org>
Date:   Wed Jan 8 21:25:19 2014 +1100

    KVM: PPC: Book3S HV: Don't set DABR on POWER8

    POWER8 doesn't have the DABR and DABRX registers; instead it has
    new DAWR/DAWRX registers, which will be handled in a later patch.

    Signed-off-by: Michael Neuling <mikey@neuling.org>
    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_interrupts.S b/arch/powerpc/kvm/book3s_hv_interrupts.S
index 928142c..00b7ed4 100644
--- a/arch/powerpc/kvm/book3s_hv_interrupts.S
+++ b/arch/powerpc/kvm/book3s_hv_interrupts.S
@@ -57,9 +57,11 @@ BEGIN_FTR_SECTION
  std r3, HSTATE_DSCR(r13)
 END_FTR_SECTION_IFSET(CPU_FTR_ARCH_206)
 
+BEGIN_FTR_SECTION
  /* Save host DABR */
  mfspr r3, SPRN_DABR
  std r3, HSTATE_DABR(r13)
+END_FTR_SECTION_IFCLR(CPU_FTR_ARCH_207S)
 
  /* Hard-disable interrupts */
  mfmsr   r10
diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index 98ff978..00b728e 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -61,11 +61,13 @@ kvmppc_call_hv_entry:
 
  /* Back from guest - restore host state and return to caller */
 
+BEGIN_FTR_SECTION
  /* Restore host DABR and DABRX */
  ld r5,HSTATE_DABR(r13)
  li r6,7
  mtspr SPRN_DABR,r5
  mtspr SPRN_DABRX,r6
+END_FTR_SECTION_IFCLR(CPU_FTR_ARCH_207S)
 
  /* Restore SPRG3 */
  ld r3,PACA_SPRG3(r13)
@@ -287,15 +289,17 @@ kvmppc_hv_entry:
  std r0, PPC_LR_STKOFF(r1)
  stdu r1, -112(r1)
 
+BEGIN_FTR_SECTION
  /* Set partition DABR */
  /* Do this before re-enabling PMU to avoid P7 DABR corruption bug */
  li r5,3
  ld r6,VCPU_DABR(r4)
  mtspr SPRN_DABRX,r5
  mtspr SPRN_DABR,r6
-BEGIN_FTR_SECTION
+ BEGIN_FTR_SECTION_NESTED(89)
  isync
-END_FTR_SECTION_IFSET(CPU_FTR_ARCH_206)
+ END_FTR_SECTION_NESTED(CPU_FTR_ARCH_206, CPU_FTR_ARCH_206, 89)
+END_FTR_SECTION_IFCLR(CPU_FTR_ARCH_207S)
 
  /* Load guest PMU registers */
  /* R4 is live here (vcpu pointer) */
@@ -1612,6 +1616,9 @@ ignore_hdec:
  b fast_guest_return
 
 _GLOBAL(kvmppc_h_set_dabr)
+BEGIN_FTR_SECTION
+ b 2f
+END_FTR_SECTION_IFSET(CPU_FTR_ARCH_207S)
  std r4,VCPU_DABR(r3)
  /* Work around P7 bug where DABR can get corrupted on mtspr */
 1: mtspr SPRN_DABR,r4
@@ -1619,7 +1626,7 @@ _GLOBAL(kvmppc_h_set_dabr)
  cmpd r4, r5
  bne 1b
  isync
- li r3,0
+2: li r3,0
  blr
 
 _GLOBAL(kvmppc_h_cede)
-- 
1.7.1